pkg/http/api_directmessaging.go

```diff
@@ -194,8 +194,22 @@ func (a *api) onDirectMessage(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	// Handle errors; successful operations are already complete
+	// If success is true, it means that headers have already been sent, so we can't send the error to the user, because:
+	// headers cannot be re-sent, and adding to response body may cause corrupted data to be sent
+	if success.Load() {
+		// Use Warn log here because it's the only way users are notified of the error
+		log.Warnf("HTTP service invocation failed to complete with error: %v", err)
+
+		// Do nothing else, as at least some data was already sent to the client
```

Comment:
Does it mean the user will get a successful response with incomplete data?

Reply:
No, if the response was successful, the data is already sent to the client within the policy function.

If we are here, it means that the request failed. There are 2 possible options here:

1. The request failed without ever sending any data to the user. In this case, down below we send a response to the client with a HTTP status code indicating an error, and the error details in the body
2. The request failed but at least a partial response has already been sent to the client. This is what we've observed when working on #7270 (which was a bug in Dapr now fixed) and now in bugs reported by 2 separate users (which seem to be due to the gRPC connection between sidecars failing - we are investigating that separately). In this case, we were here trying to re-set the status code (pointlessly since the headers have already been flushed), and more importantly we were appending the error message to the response body, which meant we were sending a "corrupted" response to the caller.

The main change here is recognizing that if some data has already been sent to the client, we should abort the request without sending _more_ data.

At the same time, we were never logging errors. Now errors are logged with debug level if they are also sent to the client in the body; if not, they are logged as warnings.